code/controllers/Processes/shuttles.dm

```diff
@@ -0,0 +1,337 @@
+		if(P.name == "HEADER") continue		// To filter out group headers
+		supply_packs["[P.type]"] = P
+	initial_move()
+
+/datum/controller/process/shuttle/doWork()
+	points += points_per_decisecond * schedule_interval
+	for(var/thing in mobile)
+		if(thing)
+			var/obj/docking_port/mobile/P = thing
+			P.check()
+			continue
+		mobile.Remove(thing)
+
+/datum/controller/process/shuttle/proc/initialize_docks()
+	for(var/obj/docking_port/D in world)
+		D.register()
+
+
+/datum/controller/process/shuttle/proc/getShuttle(id)
+	for(var/obj/docking_port/mobile/M in mobile)
+		if(M.id == id)
+			return M
+	WARNING("couldn't find shuttle with id: [id]")
+
+/datum/controller/process/shuttle/proc/getDock(id)
+	for(var/obj/docking_port/stationary/S in stationary)
+		if(S.id == id)
+			return S
+	WARNING("couldn't find dock with id: [id]")
+
+/datum/controller/process/shuttle/proc/requestEvac(mob/user, call_reason)
+	if(!emergency)
+		throw EXCEPTION("requestEvac(): There is no emergency shuttle! The game will be unresolvable. This is likely due to a mapping error")
+		return
+
+
+	switch(emergency.mode)
+		if(SHUTTLE_RECALL)
+			user << "The emergency shuttle may not be called while returning to Centcom."
+			return
+		if(SHUTTLE_CALL)
+			user << "The emergency shuttle is already on its way."
+			return
+		if(SHUTTLE_DOCKED)
+			user << "The emergency shuttle is already here."
+			return
+		if(SHUTTLE_ESCAPE)
+			user << "The emergency shuttle is moving away to a safe distance."
+			return
+		if(SHUTTLE_STRANDED)
+			user << "The emergency shuttle has been disabled by Centcom."
+			return
+
+	call_reason = trim(html_encode(call_reason))
+
+	if(length(call_reason) < CALL_SHUTTLE_REASON_LENGTH)
+		user << "You must provide a reason."
+		return
+
+	var/area/signal_origin = get_area(user)
+	var/emergency_reason = "\nNature of emergency:\n\n[call_reason]"
+	if(seclevel2num(get_security_level()) == SEC_LEVEL_RED) // There is a serious threat we gotta move no time to give them five minutes.
+		emergency.request(null, 0.5, signal_origin, html_decode(emergency_reason), 1)
+	else
+		emergency.request(null, 1, signal_origin, html_decode(emergency_reason), 0)
+
+	log_game("[key_name(user)] has called the shuttle.")
+	message_admins("[key_name_admin(user)] has called the shuttle.")
+
+	return
+
+/datum/controller/process/shuttle/proc/cancelEvac(mob/user)
+	if(canRecall())
+		emergency.cancel(get_area(user))
+		log_game("[key_name(user)] has recalled the shuttle.")
+		message_admins("[key_name_admin(user)] has recalled the shuttle.")
+		return 1
+
+/datum/controller/process/shuttle/proc/canRecall()
+	if(emergency.mode != SHUTTLE_CALL)
+		return
+	if(ticker.mode.name == "meteor")
+		return
+	if(seclevel2num(get_security_level()) == SEC_LEVEL_RED)
+		if(emergency.timeLeft(1) < emergencyCallTime * 0.25)
+			return
+	else
+		if(emergency.timeLeft(1) < emergencyCallTime * 0.5)
+			return
+	return 1
+
+/datum/controller/process/shuttle/proc/autoEvac()
```

Comment:
Thank you so much for this; this is a really nice little feature for when shit really hits the fan.